socks5.c

```diff
@@ -0,0 +1,337 @@
+    if(dill_slow(!mem)) {err = EINVAL; goto error;}
+    /* Check whether underlying socket is a bytestream. */
+    if(dill_slow(!dill_hquery(s, dill_bsock_type))) {err = errno; goto error;}
+    /* Take the ownership of the underlying socket. */
+    s = dill_hown(s);
+    if(dill_slow(s < 0)) {err = errno; goto error;}
+    /* Create the object. */
+    obj->hvfs.query = dill_socks5_hquery;
+    obj->hvfs.close = dill_socks5_hclose;
+    obj->bvfs.bsendl = dill_socks5_bsendl;
+    obj->bvfs.brecvl = dill_socks5_brecvl;
+    obj->u = s;
+    obj->mem = 1;
+    obj->server = 0;
+    obj->connect = 0;
+    /* Create the handle. */
+    int h = dill_hmake(&obj->hvfs);
+    if(dill_slow(h < 0)) {err = errno; goto error;}
+    err = dill_socks5_clientauth(h, username, password, deadline);
+    if (err) {s = h; err = errno; goto error;}
+    return h;
+error:
+    if(s >= 0) dill_hclose(s);
+    errno = err;
+    return -1;
+}
+
+int dill_socks5_attach_client(int s, const char *username,
+      const char *password, int64_t deadline) {
+    int err;
+    struct dill_socks5_sock *obj = malloc(sizeof(struct dill_socks5_sock));
+    if(dill_slow(!obj)) {err = ENOMEM; goto error1;}
+    s = dill_socks5_attach_client_mem(s, username, password,
+        (struct dill_socks5_storage*)obj, deadline);
+    if(dill_slow(s < 0)) {err = errno; goto error2;}
+    obj->mem = 0;
+    return s;
+error2:
+    free(obj);
+error1:
+    if(s >= 0) dill_hclose(s);
+    errno = err;
+    return -1;
+}
+
+int dill_socks5_connect(int s, char *addr, int port, int64_t deadline) {
+    // validate socket type
+    struct dill_socks5_sock *obj = dill_hquery(s, dill_socks5_type);
+    if(dill_slow(!obj)) return -1;
+    if(dill_slow(obj->server)) {errno = EPROTO; return -1;}
+    if(dill_slow(obj->connect)) {errno = EISCONN; return -1;}
+    // validate input
+    if(dill_slow(!addr)) {errno = EINVAL; return -1;}
+    if(dill_slow(strlen(addr) > 255)) {errno = EINVAL; return -1;}
+    if((port <= 0) || (port>65535)) {errno = EINVAL; return -1;}
+    // largest possible connect request = 255 chars for name + 7 bytes for
+    // VER, CMD, RSV, ATYP, ALEN, PORT[2]
+    uint8_t conn[262];
+    size_t conn_len;
+    conn[0] = 0x05; // VER = 5
+    conn[1] = 0x01; // CMD = CONNECT
+    conn[2] = 0x00; // RSV
+    struct in_addr ina4;
+    struct in6_addr ina6;
+    if (inet_pton(AF_INET, addr, &ina4) == 1) {
+        // IPv4 Address, now stored in ina4, Network Byte Order
+        conn[3] = 0x01; // ATYP = IPV4
+        memcpy((void *)(conn + 4), (void *)&ina4, 4);
+        conn[8] = port >> 8;
+        conn[9] = port & 0xFF;
+        conn_len = 10;
+    } else if (inet_pton(AF_INET6, addr, &ina6) == 1) {
+        // IPv6 Address, now stored in ina6, Network Byte Order
+        conn[3] = 0x04; // ATYP = IPV6
+        memcpy((void *)(conn + 4), (void *)&ina6, 16);
+        conn[20] = port >> 8;
+        conn[21] = port & 0xFF;
+        conn_len = 22;
+   } else {
```

Comment:
I see. In this case we want the proxy to do DNS query for us, correct? What about using a different variant of attach function for that kind of stuff?

Alternatively, we may want to save the user from shooting themselves in the leg and disallowing DNS queries on the user side altogether. If that's the case, don't even try to convert the string to IP address and simply pass the string to the server.

Reply:
The protocol has three options for an address. Either a literal IPv4 or IPv6 address or a name (to be resolved by the proxy). In the case of TOR at least, the name resolution needs to take place in the proxy as the client won't be able to resolve "_____.onion" addresses. 

In practice SOCKS5 traffic is sent in plaintext, so we're not likely saving anybody's leg, but having got this far I'm frankly surprised there isn't a standard option to layer SOCKS5 over TLS. 

Passing the IP address as a string isn't guaranteed to work. SOCKS5 expects these in binary form.

Reply:
I like the two functions approach. Either explicitly an IP(v4,v6) or explicitly a Name.

(also should simplify the code, which is often an indication that you're making the right choice)